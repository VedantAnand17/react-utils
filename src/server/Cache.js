/**
 * Implements the static cache.
 */
export default class Cache {
  constructor(maxSize) {
    this.private = {
      items: {},
      maxSize,
      size: 0,
    };
  }

  /**
   * Cache lookup.
   * @param {string} key Item key to look for.
   * @param {number} [maxage=Number.MAX_VALUE] Optional. The maximum age of
   *  cached item to serve. Default to infinite.
   * @returns {string} Cached item, or null if the item is absent in cache,
   *  or stale.
   */
  get({ key, maxage = Number.MAX_VALUE }) {
    const item = this.private.items[key];
    return item && Date.now() - item.timestamp < maxage ? item.data : null;
  }

  /**
   * Adds item to cache.
   * @param {string} data Item to add.
   * @param {string} key Key to store the item at.
   */
  add(data, key) {
    const p = this.private;
    const old = p.items[key];
    if (old) p.size -= old.data.length;
    p.items[key] = { data, timestamp: Date.now() };
    p.size += data.length;
    if (p.size > p.maxSize) {
      const items = Object.entries(p.items);
      items.sort((a, b) => a[1].timestamp - b[1].timestamp);
      for (let i = 0; i < items.length; ++i) {
        const item = items[i];
        delete p.items[item[0]];
        p.size -= item[1].data.length;
        if (p.size < p.maxSize / 2) break;
      }
    }
  }
}
