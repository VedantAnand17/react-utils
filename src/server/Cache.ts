/**
 * Implements the static cache.
 */
export default class Cache<DatumT> {
  private: {
    items: {
      [key: string]: {
        data: DatumT;
        size: number;
        timestamp: number;
      }
    };
    maxSize: number;
    size: number;
  };

  constructor(maxSize: number) {
    this.private = {
      items: {},
      maxSize,
      size: 0,
    };
  }

  /**
   * Cache lookup.
   * @param key Item key to look for.
   * @param [maxage=Number.MAX_VALUE] Optional. The maximum age of
   *  cached item to serve. Default to infinite.
   * @returns Cached item, or null if the item is absent in cache,
   *  or stale.
   */
  get({
    key,
    maxage = Number.MAX_VALUE,
  }: {
    key: string;
    maxage?: number;
  }): DatumT | null {
    const item = this.private.items[key];
    return item && Date.now() - item.timestamp < maxage ? item.data : null;
  }

  /**
   * Adds item to cache.
   * @ignore
   * @param data Item to add.
   * @param key Key to store the item at.
   * @param size Byte size of the item.
   */
  add(data: DatumT, key: string, size: number) {
    const p = this.private;
    const old = p.items[key];
    if (old) p.size -= old.size;
    p.items[key] = { data, size, timestamp: Date.now() };
    p.size += size;
    if (p.size > p.maxSize) {
      const items = Object.entries(p.items);
      items.sort((a, b) => a[1].timestamp - b[1].timestamp);
      for (let i = 0; i < items.length; ++i) {
        const [itemKey, item] = items[i];
        delete p.items[itemKey];
        p.size -= item.size;
        if (p.size < p.maxSize / 2) break;
      }
    }
  }
}
