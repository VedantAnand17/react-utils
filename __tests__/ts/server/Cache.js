/**
 * Testing Cache object.
 */

import mockdate from 'mockdate';
import Cache from 'server/Cache';

mockdate.set('2021-04-05Z');

it('works as expected', () => {
  const cache = new Cache(9);
  expect(cache).toMatchSnapshot();
  cache.add('123', 'A');
  mockdate.set(Date.now() + 1000);
  cache.add('123456', 'B');
  mockdate.set(Date.now() + 1000);
  expect(cache).toMatchSnapshot();
  cache.add('ab', 'A');
  mockdate.set(Date.now() + 1000);
  expect(cache).toMatchSnapshot();
  expect(cache.get({ key: 'C' })).toBe(null);
  expect(cache.get({ key: 'A' })).toBe('ab');
  expect(cache.get({ key: 'A', maxage: 0 })).toBe(null);
  cache.add('21', 'C');
  expect(cache).toMatchSnapshot();
});
