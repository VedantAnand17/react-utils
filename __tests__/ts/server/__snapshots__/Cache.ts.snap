// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works as expected 1`] = `
Cache {
  "private": {
    "items": {},
    "maxSize": 18,
    "size": 0,
  },
}
`;

exports[`works as expected 2`] = `
Cache {
  "private": {
    "items": {
      "A": {
        "data": "123",
        "size": 6,
        "timestamp": 1617580800000,
      },
      "B": {
        "data": "123456",
        "size": 12,
        "timestamp": 1617580801000,
      },
    },
    "maxSize": 18,
    "size": 18,
  },
}
`;

exports[`works as expected 3`] = `
Cache {
  "private": {
    "items": {
      "A": {
        "data": "ab",
        "size": 4,
        "timestamp": 1617580802000,
      },
      "B": {
        "data": "123456",
        "size": 12,
        "timestamp": 1617580801000,
      },
    },
    "maxSize": 18,
    "size": 16,
  },
}
`;

exports[`works as expected 4`] = `
Cache {
  "private": {
    "items": {
      "A": {
        "data": "ab",
        "size": 4,
        "timestamp": 1617580802000,
      },
      "C": {
        "data": "21",
        "size": 4,
        "timestamp": 1617580803000,
      },
    },
    "maxSize": 18,
    "size": 8,
  },
}
`;
